blockchain/indexers/common.go

```diff
@@ -173,3 +182,900 @@ func interruptRequested(ctx context.Context) bool {
+
+	c             chan *IndexNtfn
+	subscriptions map[string]*IndexSubscription
+	mtx           sync.Mutex
+	ctx           context.Context
+	cancel        context.CancelFunc
+}
+
+// NewIndexSubscriber creates a new index subscriber. It also starts the
+// handler for incoming index update subscriptions.
+func NewIndexSubscriber(sCtx context.Context) *IndexSubscriber {
+	ctx, cancel := context.WithCancel(context.Background())
+	s := &IndexSubscriber{
+		c:             make(chan *IndexNtfn, bufferSize),
+		subscriptions: make(map[string]*IndexSubscription),
+		ctx:           ctx,
+		cancel:        cancel,
+	}
+
+	go s.run(ctx)
+
+	return s
+}
+
+// Subscribe subscribes an index for updates.  The returned index subscription
+// has functions to retrieve a channel that produces a stream of index updates
+// and to stop the stream when the caller no longer wishes to receive updates.
+func (s *IndexSubscriber) Subscribe(index Indexer, prerequisite string) (*IndexSubscription, error) {
+	sub := newIndexSubscription(s, index, prerequisite)
+
+	// If the subscription has a prequisite, find it and set the subscription
+	// as a dependency.
+	if prerequisite != noPrereqs {
+		s.mtx.Lock()
+		prereq, ok := s.subscriptions[prerequisite]
+		s.mtx.Unlock()
+		if !ok {
+			return nil, fmt.Errorf("no subscription found with id %s", prerequisite)
+		}
+
+		prereq.mtx.Lock()
+		defer prereq.mtx.Unlock()
+
+		if prereq.dependent != nil {
+			return nil, fmt.Errorf("%s already has a dependent set: %s",
+				prereq.id, prereq.dependent.id)
+		}
+
+		prereq.dependent = sub
+		atomic.AddUint32(&s.subscribers, 1)
+
+		return sub, nil
+	}
+
+	// If the subscription does not have a prerequisite, add it to the index
+	// subscriber's subscriptions.
+	s.mtx.Lock()
+	s.subscriptions[sub.id] = sub
+	s.mtx.Unlock()
+
+	atomic.AddUint32(&s.subscribers, 1)
+
+	return sub, nil
+}
+
+// Notify relays an index notification to subscribed indexes for processing.
+func (s *IndexSubscriber) Notify(ntfn *IndexNtfn) {
+	subscribers := atomic.LoadUint32(&s.subscribers)
+
+	// Only relay notifications when there are subscribed indexes
+	// to be notified.
+	if subscribers > 0 {
+		s.c <- ntfn
+	}
+}
+
+// CatchUp syncs all subscribed indexes to the the main chain by
+// connecting blocks from the after the lowest index tip to the current main
+// chain tip.
+//
+// This should be called after all indexes have subscribed for updates.
+func (s *IndexSubscriber) CatchUp(ctx context.Context, db database.DB, queryer ChainQueryer) error {
+	// Find the lowest tip height to catch up from among subscribed indexes.
+	bestHeight, _ := queryer.Best()
+	lowestHeight := bestHeight
+	for _, sub := range s.subscriptions {
+		tipHeight, tipHash, err := sub.idx.Tip()
+		if err != nil {
+			return err
+		}
+
+		// Ensure the index tip is on the main chain.
+		if !queryer.MainChainHasBlock(tipHash) {
+			return fmt.Errorf("%s: index tip (%s) is not on the main chain",
+				sub.idx.Name(), tipHash)
+		}
```

Comment:
I don't think we can error out in this case.  The index tip could be on a side chain block, and it still needs to recover properly.  I had to handle this case in the [UtxoCache initialization code](https://github.com/decred/dcrd/blob/8d4ac771383f254b7a68cd89aee37e2ac19de18e/blockchain/utxocache.go#L707-L880) as well, which requires disconnecting blocks back to the fork point before replaying the main chain blocks to catch up.

Reply:
The indexes currently recover to the main chain on initialization if needed via `recover`. If `Catchup` encounters an index tip not on the main chain that'd mean the recovery process failed and should error as a result.

Reply:
Got it, thanks.  I missed that that was handled separately.